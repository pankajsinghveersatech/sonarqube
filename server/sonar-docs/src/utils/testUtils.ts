/*
 * SonarQube
 * Copyright (C) 2009-2021 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { ReactWrapper, ShallowWrapper } from 'enzyme';

export function click(element: ShallowWrapper | ReactWrapper, event = {}): void {
  // `type()` returns a component constructor for a composite element and string for DOM nodes
  if (typeof element.type() === 'function') {
    element.prop<Function>('onClick')();
    // TODO find out if `root` is a public api
    // https://github.com/airbnb/enzyme/blob/master/packages/enzyme/src/ReactWrapper.js#L109
    (element as any).root().update();
  } else {
    element.simulate('click', mockEvent(event));
  }
}

export function mockEvent(overrides = {}) {
  return {
    target: { blur() {} },
    currentTarget: { blur() {} },
    preventDefault() {},
    stopPropagation() {},
    ...overrides
  } as any;
}

export async function waitAndUpdate(wrapper: ShallowWrapper<any, any> | ReactWrapper<any, any>) {
  await new Promise(setImmediate);
  wrapper.update();
}
