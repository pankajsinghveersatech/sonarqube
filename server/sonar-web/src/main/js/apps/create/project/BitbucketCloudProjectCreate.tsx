/*
 * SonarQube
 * Copyright (C) 2009-2021 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import * as React from 'react';
import { WithRouterProps } from 'react-router';
import {
  importBitbucketCloudRepository,
  searchForBitbucketCloudRepositories
} from '../../../api/alm-integrations';
import { BitbucketCloudRepository } from '../../../types/alm-integration';
import { AlmSettingsInstance } from '../../../types/alm-settings';
import BitbucketCloudProjectCreateRenderer from './BitbucketCloudProjectCreateRender';

interface Props extends Pick<WithRouterProps, 'location' | 'router'> {
  canAdmin: boolean;
  settings: AlmSettingsInstance[];
  loadingBindings: boolean;
  onProjectCreate: (projectKey: string) => void;
}

interface State {
  importingSlug?: string;
  isLastPage?: boolean;
  loading: boolean;
  loadingMore: boolean;
  projectsPaging: Omit<T.Paging, 'total'>;
  resetPat: boolean;
  repositories: BitbucketCloudRepository[];
  searching: boolean;
  searchQuery: string;
  settings: AlmSettingsInstance;
  showPersonalAccessTokenForm: boolean;
}

export const BITBUCKET_CLOUD_PROJECTS_PAGESIZE = 30;
export default class BitbucketCloudProjectCreate extends React.PureComponent<Props, State> {
  mounted = false;

  constructor(props: Props) {
    super(props);
    this.state = {
      // For now, we only handle a single instance. So we always use the first
      // one from the list.
      loading: false,
      loadingMore: false,
      resetPat: false,
      projectsPaging: { pageIndex: 1, pageSize: BITBUCKET_CLOUD_PROJECTS_PAGESIZE },
      repositories: [],
      searching: false,
      searchQuery: '',
      settings: props.settings[0],
      showPersonalAccessTokenForm: true
    };
  }

  componentDidMount() {
    this.mounted = true;
  }

  componentDidUpdate(prevProps: Props) {
    if (prevProps.settings.length === 0 && this.props.settings.length > 0) {
      this.setState({ settings: this.props.settings[0] }, () => this.fetchData());
    }
  }

  handlePersonalAccessTokenCreated = async () => {
    this.setState({ showPersonalAccessTokenForm: false });
    this.cleanUrl();
    this.setState({ loading: true });
    await this.fetchData();
    this.setState({ loading: false });
  };

  cleanUrl = () => {
    const { location, router } = this.props;
    delete location.query.resetPat;
    router.replace(location);
  };

  async fetchData(more = false) {
    const {
      settings,
      searchQuery,
      projectsPaging: { pageIndex, pageSize },
      showPersonalAccessTokenForm
    } = this.state;
    if (settings && !showPersonalAccessTokenForm) {
      const { isLastPage, repositories } = await searchForBitbucketCloudRepositories(
        settings.key,
        searchQuery,
        pageSize,
        pageIndex
      ).catch(() => {
        this.handleError();
        return { isLastPage: undefined, repositories: undefined };
      });
      if (this.mounted && isLastPage !== undefined && repositories !== undefined) {
        if (more) {
          this.setState(state => ({
            isLastPage,
            repositories: [...state.repositories, ...repositories]
          }));
        } else {
          this.setState({ isLastPage, repositories });
        }
      }
    }
  }

  handleError = () => {
    if (this.mounted) {
      this.setState({
        projectsPaging: { pageIndex: 1, pageSize: BITBUCKET_CLOUD_PROJECTS_PAGESIZE },
        repositories: [],
        resetPat: true,
        showPersonalAccessTokenForm: true
      });
    }

    return undefined;
  };

  handleSearch = (searchQuery: string) => {
    this.setState(
      {
        searching: true,
        projectsPaging: { pageIndex: 1, pageSize: BITBUCKET_CLOUD_PROJECTS_PAGESIZE },
        searchQuery
      },
      async () => {
        await this.fetchData();
        if (this.mounted) {
          this.setState({ searching: false });
        }
      }
    );
  };

  handleLoadMore = () => {
    this.setState(
      state => ({
        loadingMore: true,
        projectsPaging: {
          pageIndex: state.projectsPaging.pageIndex + 1,
          pageSize: state.projectsPaging.pageSize
        }
      }),
      async () => {
        await this.fetchData(true);
        if (this.mounted) {
          this.setState({ loadingMore: false });
        }
      }
    );
  };

  handleImport = async (repositorySlug: string) => {
    const { settings } = this.state;

    if (!settings) {
      return;
    }

    this.setState({ importingSlug: repositorySlug });

    const result = await importBitbucketCloudRepository(settings.key, repositorySlug).catch(
      () => undefined
    );

    if (this.mounted) {
      this.setState({ importingSlug: undefined });

      if (result) {
        this.props.onProjectCreate(result.project.key);
      }
    }
  };

  render() {
    const { canAdmin, loadingBindings, location } = this.props;
    const {
      importingSlug,
      isLastPage = true,
      settings,
      loading,
      loadingMore,
      repositories,
      showPersonalAccessTokenForm,
      resetPat,
      searching,
      searchQuery
    } = this.state;
    return (
      <BitbucketCloudProjectCreateRenderer
        importingSlug={importingSlug}
        isLastPage={isLastPage}
        settings={settings}
        canAdmin={canAdmin}
        loadingMore={loadingMore}
        loading={loading || loadingBindings}
        onImport={this.handleImport}
        onLoadMore={this.handleLoadMore}
        onPersonalAccessTokenCreated={this.handlePersonalAccessTokenCreated}
        onSearch={this.handleSearch}
        repositories={repositories}
        searching={searching}
        searchQuery={searchQuery}
        resetPat={resetPat || Boolean(location.query.resetPat)}
        showPersonalAccessTokenForm={
          showPersonalAccessTokenForm || Boolean(location.query.resetPat)
        }
      />
    );
  }
}
