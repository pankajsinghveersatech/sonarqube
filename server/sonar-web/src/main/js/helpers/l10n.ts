/*
 * SonarQube
 * Copyright (C) 2009-2021 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { fetchL10nBundle } from '../api/l10n';
import { L10nBundle, L10nBundleRequestParams } from '../types/l10n';
import { toNotSoISOString } from './dates';
import { get as loadFromLocalStorage, save as saveInLocalStorage } from './storage';

export type Messages = T.Dict<string>;

export const DEFAULT_LOCALE = 'en';
export const DEFAULT_MESSAGES = {
  // eslint-disable-next-line camelcase
  default_error_message: 'The request cannot be processed. Try again later.'
};

let allMessages: Messages = {};
let locale: string | undefined;

export function translate(...keys: string[]): string {
  const messageKey = keys.join('.');
  const l10nMessages = getMessages();
  if (process.env.NODE_ENV === 'development' && !l10nMessages[messageKey]) {
    // eslint-disable-next-line no-console
    console.error(`No message for: ${messageKey}`);
  }
  return l10nMessages[messageKey] || messageKey;
}

export function translateWithParameters(
  messageKey: string,
  ...parameters: Array<string | number>
): string {
  const message = getMessages()[messageKey];
  if (message) {
    return parameters
      .map(parameter => String(parameter))
      .reduce((acc, parameter, index) => acc.replace(`{${index}}`, () => parameter), message);
  }
  if (process.env.NODE_ENV === 'development') {
    // eslint-disable-next-line no-console
    console.error(`No message for: ${messageKey}`);
  }
  return `${messageKey}.${parameters.join('.')}`;
}

export function hasMessage(...keys: string[]): boolean {
  const messageKey = keys.join('.');
  return getMessages()[messageKey] != null;
}

export function getMessages() {
  if (typeof allMessages === 'undefined') {
    logWarning('L10n messages are not initialized. Use default messages.');
    return DEFAULT_MESSAGES;
  }
  return allMessages;
}

export function resetMessages(newMessages: Messages) {
  allMessages = newMessages;
}

export function getLocalizedMetricName(
  metric: { key: string; name?: string },
  short = false
): string {
  const bundleKey = `metric.${metric.key}.${short ? 'short_name' : 'name'}`;
  if (hasMessage(bundleKey)) {
    return translate(bundleKey);
  } else if (short) {
    return getLocalizedMetricName(metric);
  }
  return metric.name || metric.key;
}

export function getLocalizedCategoryMetricName(metric: { key: string; name?: string }) {
  const bundleKey = `metric.${metric.key}.extra_short_name`;
  return hasMessage(bundleKey) ? translate(bundleKey) : getLocalizedMetricName(metric, true);
}

export function getLocalizedMetricDomain(domainName: string) {
  const bundleKey = `metric_domain.${domainName}`;
  return hasMessage(bundleKey) ? translate(bundleKey) : domainName;
}

export function getCurrentLocale() {
  return locale;
}

export function resetCurrentLocale(newLocale: string) {
  locale = newLocale;
}

export function getShortMonthName(index: number) {
  const months = [
    'Jan',
    'Feb',
    'Mar',
    'Apr',
    'May',
    'Jun',
    'Jul',
    'Aug',
    'Sep',
    'Oct',
    'Nov',
    'Dec'
  ];
  return translate(months[index]);
}

export function getWeekDayName(index: number) {
  const weekdays = ['Sunday', 'Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday'];
  return weekdays[index] ? translate(weekdays[index]) : '';
}

export function getShortWeekDayName(index: number) {
  const weekdays = ['Sun', 'Mon', 'Tue', 'Wed', 'Thu', 'Fri', 'Sat'];
  return weekdays[index] ? translate(weekdays[index]) : '';
}

const L10N_BUNDLE_LS_KEY = 'l10n.bundle';

export async function loadL10nBundle() {
  const bundle = await getLatestL10nBundle().catch(() => ({
    locale: DEFAULT_LOCALE,
    messages: {}
  }));

  resetCurrentLocale(bundle.locale);
  resetMessages(bundle.messages);

  // No need to load english (default) bundle, it's coming with react-intl
  if (bundle.locale !== DEFAULT_LOCALE) {
    const [intlBundle, intl] = await Promise.all([
      import(`react-intl/locale-data/${bundle.locale}`),
      import('react-intl')
    ]);

    intl.addLocaleData(intlBundle.default);
  }

  return bundle;
}

export async function getLatestL10nBundle() {
  const browserLocale = getPreferredLanguage();
  const cachedBundle = loadL10nBundleFromLocalStorage();

  const params: L10nBundleRequestParams = {};

  if (browserLocale) {
    params.locale = browserLocale;

    if (
      cachedBundle.locale &&
      browserLocale.startsWith(cachedBundle.locale) &&
      cachedBundle.timestamp &&
      cachedBundle.messages
    ) {
      params.ts = cachedBundle.timestamp;
    }
  }

  const { effectiveLocale, messages } = await fetchL10nBundle(params).catch(response => {
    if (response && response.status === 304) {
      return {
        effectiveLocale: cachedBundle.locale || browserLocale || DEFAULT_LOCALE,
        messages: cachedBundle.messages ?? {}
      };
    }
    throw new Error(`Unexpected status code: ${response.status}`);
  });

  const bundle = {
    timestamp: toNotSoISOString(new Date()),
    locale: effectiveLocale,
    messages
  };

  saveL10nBundleToLocalStorage(bundle);

  return bundle;
}

export function getCurrentL10nBundle() {
  return loadL10nBundleFromLocalStorage();
}

function getPreferredLanguage() {
  return window.navigator.languages ? window.navigator.languages[0] : window.navigator.language;
}

function loadL10nBundleFromLocalStorage() {
  let bundle: L10nBundle;

  try {
    bundle = JSON.parse(loadFromLocalStorage(L10N_BUNDLE_LS_KEY) ?? '{}');
  } catch {
    bundle = {};
  }

  return bundle;
}

function saveL10nBundleToLocalStorage(bundle: L10nBundle) {
  saveInLocalStorage(L10N_BUNDLE_LS_KEY, JSON.stringify(bundle));
}

function logWarning(message: string) {
  if (process.env.NODE_ENV !== 'production') {
    // eslint-disable-next-line no-console
    console.warn(message);
  }
}
