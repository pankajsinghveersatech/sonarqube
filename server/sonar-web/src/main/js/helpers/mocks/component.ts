/*
 * SonarQube
 * Copyright (C) 2009-2021 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { ComponentQualifier, TreeComponent } from '../../types/component';
import { MetricKey } from '../../types/metrics';
import { mockMeasureEnhanced } from '../testMocks';

export function mockComponent(overrides: Partial<T.Component> = {}): T.Component {
  return {
    breadcrumbs: [],
    key: 'my-project',
    name: 'MyProject',
    qualifier: ComponentQualifier.Project,
    qualityGate: { isDefault: true, key: '30', name: 'Sonar way' },
    qualityProfiles: [
      {
        deleted: false,
        key: 'my-qp',
        language: 'ts',
        name: 'Sonar way'
      }
    ],
    tags: [],
    ...overrides
  };
}

export function mockTreeComponent(overrides: Partial<TreeComponent>): TreeComponent {
  return {
    key: 'my-key',
    qualifier: ComponentQualifier.Project,
    name: 'component',
    visibility: 'public',
    ...overrides
  };
}

export function mockComponentMeasure(
  file = false,
  overrides: Partial<T.ComponentMeasure> = {}
): T.ComponentMeasure {
  if (file) {
    return {
      key: 'foo:src/index.tsx',
      name: 'index.tsx',
      qualifier: ComponentQualifier.File,
      path: 'src/index.tsx',
      measures: [{ metric: MetricKey.bugs, value: '1', bestValue: false }],
      ...overrides
    };
  }
  return {
    key: 'foo',
    name: 'Foo',
    qualifier: ComponentQualifier.Project,
    measures: [{ metric: MetricKey.bugs, value: '12', bestValue: false }],
    ...overrides
  };
}

export function mockComponentMeasureEnhanced(
  overrides: Partial<T.ComponentMeasureEnhanced> = {}
): T.ComponentMeasureEnhanced {
  return {
    ...mockComponentMeasure(false, overrides as T.ComponentMeasure),
    leak: undefined,
    measures: [mockMeasureEnhanced()],
    value: undefined,
    ...overrides
  };
}
