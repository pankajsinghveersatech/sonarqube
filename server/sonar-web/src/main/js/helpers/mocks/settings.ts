/*
 * SonarQube
 * Copyright (C) 2009-2021 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { Setting, SettingCategoryDefinition, SettingWithCategory } from '../../types/settings';

export function mockDefinition(
  overrides: Partial<SettingCategoryDefinition> = {}
): SettingCategoryDefinition {
  return {
    key: 'foo',
    category: 'foo category',
    fields: [],
    options: [],
    subCategory: 'foo subCat',
    ...overrides
  };
}

export function mockSetting(overrides: Partial<Setting> = {}): Setting {
  return {
    key: 'foo',
    value: '42',
    inherited: true,
    definition: {
      key: 'foo',
      name: 'Foo setting',
      description: 'When Foo then Bar',
      type: 'INTEGER',
      options: []
    },
    ...overrides
  };
}

export function mockSettingWithCategory(
  overrides: Partial<SettingWithCategory> = {}
): SettingWithCategory {
  return {
    key: 'foo',
    value: '42',
    inherited: true,
    definition: {
      key: 'foo',
      name: 'Foo setting',
      description: 'When Foo then Bar',
      type: 'INTEGER',
      options: [],
      category: 'general',
      fields: [],
      subCategory: 'email'
    },
    ...overrides
  };
}
