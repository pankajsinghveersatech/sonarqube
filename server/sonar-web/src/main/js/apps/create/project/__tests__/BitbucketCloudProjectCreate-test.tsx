/*
 * SonarQube
 * Copyright (C) 2009-2021 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { shallow } from 'enzyme';
import * as React from 'react';
import {
  importBitbucketCloudRepository,
  searchForBitbucketCloudRepositories
} from '../../../../api/alm-integrations';
import { mockBitbucketCloudRepository } from '../../../../helpers/mocks/alm-integrations';
import { mockBitbucketCloudAlmSettingsInstance } from '../../../../helpers/mocks/alm-settings';
import { mockLocation, mockRouter } from '../../../../helpers/testMocks';
import { waitAndUpdate } from '../../../../helpers/testUtils';
import BitbucketCloudProjectCreate, {
  BITBUCKET_CLOUD_PROJECTS_PAGESIZE
} from '../BitbucketCloudProjectCreate';

jest.mock('../../../../api/alm-integrations', () => {
  const { mockProject } = jest.requireActual('../../../../helpers/mocks/projects');
  return {
    importBitbucketCloudRepository: jest
      .fn()
      .mockResolvedValue({ project: mockProject({ key: 'project-key' }) }),
    searchForBitbucketCloudRepositories: jest
      .fn()
      .mockResolvedValue({ isLastPage: true, repositories: [] }),
    checkPersonalAccessTokenIsValid: jest.fn().mockResolvedValue({ status: true }),
    setAlmPersonalAccessToken: jest.fn().mockResolvedValue(null)
  };
});

it('Should render correctly', async () => {
  const wrapper = shallowRender({ settings: [] });
  await waitAndUpdate(wrapper);
  expect(wrapper).toMatchSnapshot();
  wrapper.setProps({ settings: [mockBitbucketCloudAlmSettingsInstance()] });
  await waitAndUpdate(wrapper);
  expect(wrapper).toMatchSnapshot('Setting changeds');
});

it('Should handle app password correctly', async () => {
  const wrapper = shallowRender();

  await waitAndUpdate(wrapper);
  await wrapper.instance().handlePersonalAccessTokenCreated();
  expect(wrapper.state().showPersonalAccessTokenForm).toBe(false);
});

it('Should handle error correctly', async () => {
  (searchForBitbucketCloudRepositories as jest.Mock).mockRejectedValueOnce({});

  const wrapper = shallowRender();
  wrapper.setState({
    showPersonalAccessTokenForm: false,
    repositories: [mockBitbucketCloudRepository()],
    projectsPaging: { pageIndex: 2, pageSize: BITBUCKET_CLOUD_PROJECTS_PAGESIZE }
  });
  await wrapper.instance().handlePersonalAccessTokenCreated();
  expect(wrapper.state().repositories).toHaveLength(0);
  expect(wrapper.state().projectsPaging.pageIndex).toBe(1);
  expect(wrapper.state().showPersonalAccessTokenForm).toBe(true);
  expect(wrapper.state().resetPat).toBe(true);
});

it('Should load repository', async () => {
  (searchForBitbucketCloudRepositories as jest.Mock).mockResolvedValueOnce({
    isLastPage: true,
    repositories: [
      mockBitbucketCloudRepository(),
      mockBitbucketCloudRepository({ sqProjectKey: 'sq-key' })
    ]
  });

  const wrapper = shallowRender();
  await wrapper.instance().handlePersonalAccessTokenCreated();
  expect(wrapper.state().repositories).toHaveLength(2);
});

it('Should load more repository', async () => {
  (searchForBitbucketCloudRepositories as jest.Mock).mockResolvedValueOnce({
    isLastPage: true,
    repositories: [
      mockBitbucketCloudRepository(),
      mockBitbucketCloudRepository({ sqProjectKey: 'sq-key' })
    ]
  });

  const wrapper = shallowRender();
  wrapper.setState({ showPersonalAccessTokenForm: false, isLastPage: false });
  wrapper.instance().handleLoadMore();
  await waitAndUpdate(wrapper);
  expect(wrapper.state().repositories).toHaveLength(2);
  expect(wrapper.state().projectsPaging.pageIndex).toBe(2);
});

it('Should handle search repository', async () => {
  (searchForBitbucketCloudRepositories as jest.Mock).mockResolvedValueOnce({
    isLastPage: true,
    repositories: [
      mockBitbucketCloudRepository(),
      mockBitbucketCloudRepository({ sqProjectKey: 'sq-key' })
    ]
  });

  const wrapper = shallowRender();
  wrapper.setState({
    isLastPage: false,
    showPersonalAccessTokenForm: false,
    projectsPaging: { pageIndex: 2, pageSize: BITBUCKET_CLOUD_PROJECTS_PAGESIZE },
    repositories: [mockBitbucketCloudRepository()]
  });
  wrapper.instance().handleSearch('test');
  await waitAndUpdate(wrapper);
  expect(wrapper.state().repositories).toHaveLength(2);
  expect(wrapper.state().projectsPaging.pageIndex).toBe(1);
  expect(searchForBitbucketCloudRepositories).toHaveBeenLastCalledWith(
    'key',
    'test',
    BITBUCKET_CLOUD_PROJECTS_PAGESIZE,
    1
  );
});

it('Should import repository', async () => {
  const onProjectCreate = jest.fn();
  const wrapper = shallowRender({ onProjectCreate });
  wrapper.setState({
    isLastPage: false,
    showPersonalAccessTokenForm: false,
    projectsPaging: { pageIndex: 1, pageSize: BITBUCKET_CLOUD_PROJECTS_PAGESIZE },
    repositories: [mockBitbucketCloudRepository({ slug: 'slug-test' })]
  });
  await wrapper.instance().handleImport('slug-test');
  expect(importBitbucketCloudRepository).toHaveBeenCalledWith('key', 'slug-test');
  expect(onProjectCreate).toHaveBeenCalledWith('project-key');
});

it('Should behave correctly when import fail', async () => {
  (importBitbucketCloudRepository as jest.Mock).mockRejectedValueOnce({});
  const onProjectCreate = jest.fn();
  const wrapper = shallowRender({ onProjectCreate });
  wrapper.setState({
    isLastPage: false,
    showPersonalAccessTokenForm: false,
    projectsPaging: { pageIndex: 1, pageSize: BITBUCKET_CLOUD_PROJECTS_PAGESIZE },
    repositories: [mockBitbucketCloudRepository({ slug: 'slug-test' })]
  });
  await wrapper.instance().handleImport('slug-test');
  expect(importBitbucketCloudRepository).toHaveBeenCalledWith('key', 'slug-test');
  expect(onProjectCreate).not.toHaveBeenCalled();
});

function shallowRender(props?: Partial<BitbucketCloudProjectCreate['props']>) {
  return shallow<BitbucketCloudProjectCreate>(
    <BitbucketCloudProjectCreate
      onProjectCreate={jest.fn()}
      loadingBindings={false}
      location={mockLocation()}
      canAdmin={true}
      router={mockRouter()}
      settings={[mockBitbucketCloudAlmSettingsInstance()]}
      {...props}
    />
  );
}
