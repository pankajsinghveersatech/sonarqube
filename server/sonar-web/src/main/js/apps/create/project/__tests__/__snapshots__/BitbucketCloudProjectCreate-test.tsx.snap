// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render correctly 1`] = `
<BitbucketCloudProjectCreateRenderer
  canAdmin={true}
  isLastPage={true}
  loading={false}
  loadingMore={false}
  onImport={[Function]}
  onLoadMore={[Function]}
  onPersonalAccessTokenCreated={[Function]}
  onSearch={[Function]}
  repositories={Array []}
  resetPat={false}
  searchQuery=""
  searching={false}
  showPersonalAccessTokenForm={true}
/>
`;

exports[`Should render correctly: Setting changeds 1`] = `
<BitbucketCloudProjectCreateRenderer
  canAdmin={true}
  isLastPage={true}
  loading={false}
  loadingMore={false}
  onImport={[Function]}
  onLoadMore={[Function]}
  onPersonalAccessTokenCreated={[Function]}
  onSearch={[Function]}
  repositories={Array []}
  resetPat={false}
  searchQuery=""
  searching={false}
  settings={
    Object {
      "alm": "bitbucketcloud",
      "key": "key",
    }
  }
  showPersonalAccessTokenForm={true}
/>
`;
